Fabric scatter write tracing support + additional fixes for noc event tracing

sohaibnadeemTT opened the pull request:
### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes: https://github.com/tenstorrent/tt-metal/actions/runs/16281328941
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable): https://github.com/tenstorrent/tt-metal/actions/runs/16281330777
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

Additional:
- T3K Profiler Tests: https://github.com/tenstorrent/tt-metal/actions/runs/16281336368/job/45971904408

sohaibnadeemTT commented on tt_metal/tools/profiler/event_metadata.hpp:
```diff
@@ -42,6 +42,23 @@ struct alignas(uint64_t) KernelProfilerNocEventMetadata {
+        FabricPacketType routing_fields_type;
+    };
+
+    // struct FabricNoCScatterEventv2 {
```
I could have used something like `FabricNoCScatterEventv2` or even just add a `payload_size` field to `FabricNoCEvent` (there should be just enough space for it if we pack the struct?) but I opted for using a different struct in case scatter write changes to be more than two chunks.

bgrady-tt replied:
I think this is the right call.

bgrady-tt commented on tt_metal/impl/profiler/profiler.cpp:
```diff
@@ -1206,23 +1237,49 @@ void DeviceProfiler::serializeJsonNocTraces(
         for (size_t i = 0; i < events.size(); /* manual increment */) {
-            const auto& current_event = events[i];
+            // log_info(tt::LogMetal, "------------event index: {}----------", i);
+            // log_info(tt::LogMetal, "{}", events[i].dump(2));
```
remove commented log statements 🙏

bgrady-tt commented on tt_metal/impl/profiler/profiler.cpp:
```diff
@@ -1206,23 +1237,49 @@ void DeviceProfiler::serializeJsonNocTraces(
         }
     };
 
+    log_info(tt::LogMetal, "------------processing profiling events----------");
```
remove 😄

bgrady-tt reviewed (approved):
LGTM 👍

Commit:
Fabric scatter write tracing support

Commit:
Miscellaneous fixes for NoC event tracing:
- change routing info dump to be at profiler init
- pass correct device to npe from tracy (using cluster type in
  topology.json)
- fix address decode for recording noc multicast

mo-tenstorrent commented on ttnn/api/tools/profiler/op_profiler.hpp:
```diff
@@ -441,7 +441,7 @@ inline std::string op_meta_data_serialized_json(
         j["kernel_info"] = get_kernels_json(device_id, program);
 
         auto opname = j["op_code"].template get<std::string>();
-        runtime_id_to_opname_.insert({device_id, program.get_runtime_id()}, opname);
```
I would leave this as it was and drop the operation_id argument altogether as part of another PR

sohaibnadeemTT replied:
The operation_id comes from tt::tt_metal::detail::EncodePerDeviceProgramID and is the same host runtime id the profiler is collecting so for consistency I would have to get rid of that entire change (tt::tt_metal::detail::EncodePerDeviceProgramID and all uses in metal and ttnn). Is that what you mean?

mo-tenstorrent replied:
Program runtime_id and operation_id actually require to be the same and since program is passed here, we should not need to pass operation_id.

sohaibnadeemTT replied:
It isn't the same anymore, the operation id comes from EncodePerDeviceProgramID which encodes the  runtime id and device id in a single 32 bit int:

EncodePerDeviceProgramID:
<img width="905" height="168" alt="image" src="https://github.com/user-attachments/assets/cb05e9d8-3885-4600-9c30-240e0f69d881" />

structure specified in tt-metal/tt_metal/hw/inc/dev_msgs.h:
<img width="740" height="140" alt="image" src="https://github.com/user-attachments/assets/411c0dc5-f5af-495d-8b34-28c108711395" />

bgrady-tt reviewed (approved):
LGTM 👍

sohaibnadeemTT commented on tt_metal/impl/profiler/profiler.cpp:
```diff
@@ -1484,13 +1534,22 @@ void DeviceProfiler::dumpResults(
     if (state == ProfilerDumpState::NORMAL && rtoptions.get_profiler_noc_events_enabled()) {
         serializeJsonNocTraces(noc_trace_json_log, rpt_path, device_id, routing_lookup);
```
Yeah, I needed to get this merged today, but noted and will fix in the next PR for sure :).
